drivers/tee/optee/optee.c

```diff
@@ -286,19 +305,125 @@ static void handle_cmd_get_time(const struct device *dev, struct optee_msg_arg *
 	arg->ret = TEEC_SUCCESS;
 }
 
-static void handle_cmd_notify(const struct device *dev, struct optee_msg_arg *arg)
+/* This should be called under notif_lock */
+static inline bool key_is_pending(struct optee_driver_data *data, uint32_t key)
 {
+	struct optee_notify *iter;
+
+	SYS_DLIST_FOR_EACH_CONTAINER(&data->notif, iter, node) {
+		if (iter->key == key) {
+			k_sem_give(&iter->wait);
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int optee_notif_send(const struct device *dev, uint32_t key)
+{
+	struct optee_driver_data *data = dev->data;
+	k_spinlock_key_t sp_key;
+
+	if (key > CONFIG_OPTEE_MAX_NOTIF) {
+		return -EINVAL;
+	}
+
+	sp_key = k_spin_lock(&data->notif_lock);
+	if (!key_is_pending(data, key)) {
+		/* If nobody is waiting for key - set bit in the bitmap */
+		sys_bitarray_set_bit(&notif_bitmap, key);
+	}
+	k_spin_unlock(&data->notif_lock, sp_key);
+
+	return 0;
+}
+
+static int optee_notif_wait(const struct device *dev, uint32_t key)
+{
+	int rc = 0;
+	struct optee_driver_data *data = dev->data;
+	struct optee_notify *entry;
+	k_spinlock_key_t sp_key;
+	int prev_val;
+
+	if (key > CONFIG_OPTEE_MAX_NOTIF)
+		return -EINVAL;
+
+	entry = k_malloc(sizeof(*entry));
+	if (!entry) {
+		return -ENOMEM;
+	}
+
+	k_sem_init(&entry->wait, 0, 1);
+	entry->key = key;
+
+	sp_key = k_spin_lock(&data->notif_lock);
+
+	/*
+	 * If notif bit was set then SEND command was already received.
+	 * Skipping wait.
+	 */
+	rc = sys_bitarray_test_and_clear_bit(&notif_bitmap, key, &prev_val);
+	if (rc || prev_val) {
+		goto out;
+	}
+
 	/*
-	 * TODO: Implement wait for completion based on zephyr workq
-	 * Just sleep for 10 msec for now.
+	 * If key is already registred, then skip.
 	 */
-	k_sleep(K_MSEC(10));
+	if (key_is_pending(data, key)) {
+		rc = -EBUSY;
+		goto out;
+	}
+
+	sys_dlist_append(&data->notif, &entry->node);
+
+	k_spin_unlock(&data->notif_lock, sp_key);
+	k_sem_take(&entry->wait, K_FOREVER);
+	sp_key = k_spin_lock(&data->notif_lock);
+
+	sys_dlist_remove(&entry->node);
+out:
+	k_spin_unlock(&data->notif_lock, sp_key);
+
+	k_free(entry);
+
+	return rc;
+}
+
+static void handle_cmd_notify(const struct device *dev, struct optee_msg_arg *arg)
+{
+	if (!check_param_input(arg)) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	switch (arg->params[0].u.value.a) {
+	case OPTEE_RPC_NOTIFICATION_SEND:
+		if (optee_notif_send(dev, arg->params[0].u.value.b)) {
+			goto err;
+		}
+		break;
+	case OPTEE_RPC_NOTIFICATION_WAIT:
+		if (optee_notif_wait(dev, arg->params[0].u.value.b)) {
+			goto err;
+		}
+		break;
+	default:
+		goto err;
+	}
+
 	arg->ret = TEEC_SUCCESS;
+	return;
+
+err:
+	arg->ret = TEEC_ERROR_BAD_PARAMETERS;
 }
 
 static void handle_cmd_wait(const struct device *dev, struct optee_msg_arg *arg)
 {
-	if (check_param_input(arg)) {
+	if (!check_param_input(arg)) {
```

Comment:
This is the wrong commit for this change

Reply:
fixed